lib/response.js

```diff
@@ -766,13 +766,29 @@ internals.Response.Peek = class extends Stream.Transform {
         super();
 
         this._podium = podium;
-        this.on('finish', () => podium.emit('finish'));
     }
 
-    _transform(chunk, encoding, callback) {
+    async _transform(chunk, encoding, callback) {
 
-        this._podium.emit('peek', [chunk, encoding]);
-        this.push(chunk, encoding);
-        callback();
+        try {
+            await this._podium.emit('peek', [chunk, encoding]);
+
+            callback(null, chunk);
+        }
+        catch (err) {
+            callback(err);
+        }
+    }
+
+    async _flush(callback) {
```

Comment:
You should not abuse the `_flush()` to handler to handle a `finish` event. If you need to signal a stream error, call `this.destroy(err)`.

Reply:
If what you mean is doing the following, then it won't work because the the error will be thrown asynchronously, and even possibly after the stream was already destroyed.

```js
    constructor(podium) {

        super();

        this._podium = podium;
        this.on('finish', async () => {

            try {
                await this._podium.emit('finish');
            } catch (err) {
                this.destroy(err);
            }
        });
    }
```

Reading node's stream doc I realize that `_final` is a better place to perform this.

However I am not sure I follow you regarding `this.destroy(err)`. [the doc](https://nodejs.org/api/stream.html#stream_writable_final_callback) seem to specify that `callback` should always be called, with an error if need be.

Reply:
Reading node's stream doc I realize that `_final` is probably better place to perform this.

Reply:
no sure why you'd want to call `this.destroy(err)` directly.
It is my understanding that `this.destroy(err)` should only be performed if the `autoDestroy` option is set, which is what `callback(err)` is [actually doing](https://github.com/nodejs/node/blob/9c62e0e384e2ea5a45a4011f31c97be60f2a0737/lib/internal/streams/writable.js#L651-L657).

Reply:
Placing it in `_final()` seems better.